File: Java/OOP February - April 2021/02. Encapsulation/E05FootballTeamGenerator/Team.java
package E05FootballTeamGenerator;

import java.util.ArrayList;
import java.util.List;

public class Team {
    private String name;
    private List<Player> players;

    public Team(String name) {
        this.setName(name);
        this.players = new ArrayList<>();
    }

    private void setName(String teamName) {
        if (teamName == null || teamName.trim().isEmpty()) {
            throw new IllegalArgumentException("A name should not be empty.");
        }
        this.name = teamName;
    }

    public String getName() {
        return this.name;
    }

    public void addPlayer(Player player) {
        if (this.name == null) {

        }
        this.players.add(player);
    }

    public void removePlayer(String playerName) {
        Player player = this.players.stream().filter(p -> p.getName().equals(playerName)).findFirst().orElse(null);

        if (player == null) {
            throw new IllegalArgumentException(String.format("Player %s is not in %s team.", playerName, this.name));
        } else {
            this.players.remove(player);
        }
    }

    public double getRating() {
        double sumOfRatings = 0;
        for (Player player : players) {
            sumOfRatings += player.overallSkillLevel();
        }

        return sumOfRatings / this.players.size();
    }

}

File: C#/OOP/02_Encapsulation_Exercise/FootballTeamGenerator/Model/Team.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace FootballTeamGenerator.Model
{
    public class Team
    {
        private string name;
        private List<Player> players;

        public Team(string name)
        {
            Name = name;

            players = new List<Player>();
        }

        public string Name
        {
            get { return this.name; }
            private set
            {
                if (string.IsNullOrWhiteSpace(value))
                {
                    throw new ArgumentException("A name should not be empty.");
                }

                this.name = value;
            }
        }

        public double Rating
        {
            get
            {
                if (players.Any())
                {
                    return players.Average(p => p.Stats());
                }

                return 0;
            }
        }

        public void AddPlayer(Player player)
        {
            players.Add(player);
        }

        public void RemovePlayer(string playerName)
        {
            Player player = players.FirstOrDefault(p => p.Name == playerName);

            if (player == null)
            {
                throw new ArgumentException($"Player {playerName} is not in {Name} team.");
            }

            players.Remove(player);
        }
    }
}
